### PR TITLE
RoR Project Third Completion

### PR DESCRIPTION
**Pull request by ometman**
I am completed the Budget App project with the following:

**Interactions**
    - Splash screen
    - A simple page with the name of the app and links to the sign up and log in pages.
   
**Created the Sign up and log in pages**
    - A user can register in the app with full name, email and password (all mandatory).
    - The user can log into the app using email and password.
    - If the user is not logged in, they can't access pages that require the user to be logged in

**Home or Categories page**
    - When a user logs in, they are presented with the categories page.
    - For each category, the user can see their name, icon and the total amount of all transactions belonging to that category.
    - Clicking a category item, it navigates to the transactions page for that category.
    - There is also a button to "add a new category" at the bottom that brings the user to the page to create a new category.

**Transactions page**
    - For a given category, the list of transactions is presented, ordered by the most recent.
    - At the top of the page tis the total amount for the category.
    - There is a button "add a new transaction" at the bottom that brings the user to the page to create a new transaction.
    - When the user clicks on the "Back" button (<), the user navigates to the home page.

**The "Add a new category" page**
    - The user fills out a form to create a new category, indicating their name and icon (both mandatory).
    - The user clicks (or taps) the "Save" button to create the new category, and is taken to the home page on success.
    - When the user clicks on the "Back" button (<), the user navigates to the home page.

**"The Add a new transaction" page**
    - The user fills out a form to create a new transaction with:
        - [ ] name (mandatory)
        - [ ] - amount (mandatory)
        - [ ] - categories (mandatory at least one)
    - Tapping or clicking the "Save" button creates a new transaction, and redirects to the transactions page for that category.
    - When the user clicks on the "Back" button (<), the user navigates to the transactions page for that category.

**Testing requirements**
    - Unit and integrated tests were written for all the most important components of the application.

**Technically**
    - Postgres is used as database
    - View the schema: https://github.com/microverseinc/curriculum-rails/blob/main/capstone/images/erd_diagram.png
    - The Design is by **Gregoire Vella** of Behance https://www.behance.net/gregoirevella
    - Devise and Cancancan is used for authentication and authorization respectively.
    - All users are validated and data is private.
    - The template engine is '.erb'
    - Best practices checked with linters and passed

### CODE REVIEW

File: app/views/categories/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            <div class="category-container">
+              <%= image_tag url_for(category.icon.variant(resize: "100x100")) if category.icon.attached? %>
+              <div class="cat-name-date">
+                <h2 class="cat-name"><%= category.name %></h2>
+                <p class="cat-date"><%= category.created_at %></p>
+              </div>
+              <span class="cat-total-amt"><%= category.deals.sum(:amount) %></span>
+            </div>
+          <% end %>
+        <% end %>
```
Comment (thecodechaser) [inline app/views/categories/index.html.erb]: - [ ] I'm not able to see the uploaded icon of categories, They're broken. Please make sure to fix this and categories icons should display on the screen.

image

File: app/views/deals/_form.html.erb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        <p>
+        <%= form.collection_select(:category_id, current_user.categories, :id, :name, { include_hidden: false }, { multiple: false }) %>
+        </p>
+      </div>
+
+    <% end %>
+  </div>
+  <%= link_to "Back to Transactions", user_category_path(current_user, @category) %>
+<div class="category-actions">
+<%= link_to "SAVE", 'javascript:void(0)', onclick: 'getElementById("new-transaction-save-btn").submit();', class: 'new-category-btn' %>
```
Comment (thecodechaser) [inline app/views/deals/_form.html.erb:26]: - [ ] After creating a transaction successfully users are redirected to a blank page. Please make sure to fix this issue and after successfully creating transactions users should be redirected to the transactions page for that category as requested in the project requirements.

image

File: spec/features/devise/confirmations_new_spec.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    visit new_user_confirmation_path
+
+    within('.signup-container') do
+      expect(page).to have_field('Email', type: 'email')
+      expect(page).to have_selector('input[type="submit"].signup-submit', visible: :hidden)
+    end
+
+    fill_in 'Email', with: 'user@example.com'
+    execute_script("document.getElementById('confirmation-form').submit();")
```
Comment (thecodechaser) [inline spec/features/devise/confirmations_new_spec.rb:16]: - [ ] The below tests are failing, Please make sure that all tests pass without any errors.

image

File: README.md [hunk 4]
```diff
@@ -1,2 +1,246 @@
...
+
+# 📖 Ometman Budget App <a name="about-project"></a>
+
+**A Demo Video: Introduction to this Project**
+--[Budget Demo](https://www.loom.com/share/7c5642f53ab6429cbbdd520937e72812?sid=3b5006c2-5d4e-4eb2-bcf0-ad366796f2f3)
+
+> The project is mobile-oriented designs for different screen sizes.
+
+- It is accessible on any screen.
+- Better view on Mobile Devices and Ipads.
```
Comment (thecodechaser) [inline README.md:43]: - [ ] Please make sure that the project is deployed and accessible online and add the live demo link in the README file as requested in the project requirements.

**Review by thecodechaser — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ 

## Hi @ometman👋, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## To Highlight 🎉
- Good job with model validations. ✔️ 
- Nice UI design. ✔️
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- No linter errors. :heavy_check_mark:
- PR has a good title and summary. ✔️



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/deals/_form.html.erb (hunk 2 above)
Comment (ometman) [reply app/views/deals/_form.html.erb:26]: Thank you. Fixed

File: spec/features/devise/confirmations_new_spec.rb (hunk 3 above)
Comment (ometman) [reply spec/features/devise/confirmations_new_spec.rb:16]: Thanks. It's fixed. All tests now pass successfully.

File: app/views/categories/index.html.erb (hunk 1 above)
Comment (topeogunleye) [inline app/views/categories/index.html.erb:22]: - [ ] As suggested by the previous code reviewer, the uploaded icons of categories are broken, please fix it.

image

File: spec/features/devise/confirmations_new_spec.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
```
Comment (topeogunleye) [inline spec/features/devise/confirmations_new_spec.rb:16]: -  [ ] The tests are still failing as seen below. please fix them:


image

File: README.md [hunk 6]
```diff
@@ -1,2 +1,246 @@
...
+- Voila! Now, go to the browser and enter: localhost:3000 or 127.0.0.1:3000
+
+### Deployment
+
+You can deploy this project using:
+
+- GitHub Pages from a specific branch (e.g. main).
+- You can also used any cloud service like gcloud or Azure.
+- A deployed version can be found here:
+ [https://proper-rivalee-microverse-student.koyeb.app/]
```
Comment (topeogunleye) [inline README.md:187]: - [ ] Please kindly know that your deployed project has an error. please fix:

image

**Review by topeogunleye — CHANGES_REQUESTED** [review]
Hi @ometman, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
*  After creating a transaction successfully users are redirected to transactions page :heavy_check_mark:

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

You can also consider:
- x
- y
- z

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @topeogunleye in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/devise/confirmations_new_spec.rb (hunk 5 above)
Comment (ometman) [reply spec/features/devise/confirmations_new_spec.rb:16]: All tests pass. Some will fail if the server is on during the test, in the Capybara environment.

File: app/views/categories/index.html.erb (hunk 1 above)
Comment (ometman) [reply app/views/categories/index.html.erb:22]: Thanks. The bug is fixed. 😄

**Pull request closed without merging**